pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -600,6 +629,38 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 		changes.KillPod = true
 	}
 
+	if pod.Spec.RestartPolicy != v1.RestartPolicyAlways {
+		onlySidecars := true
+
+		//Determine if there are any non sidecar containers that are running or need restarting
+		//if there are none, we can kill the remaining sidecars
+		for _, container := range pod.Spec.Containers {
+			containerStatus := podStatus.FindContainerStatusByName(container.Name)
+			if !container.Sidecar {
+				if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) || containerStatus.State == kubecontainer.ContainerStateRunning {
+					onlySidecars = false
+				}
+			}
+		}
+
+		//only sidecars are left so terminate them all
+		if onlySidecars {
+			for idx, container := range pod.Spec.Containers {
+				containerStatus := podStatus.FindContainerStatusByName(container.Name)
+				// we don't need to terminate non sidecars or exited sidecars
+				if container.Sidecar && containerStatus.State == kubecontainer.ContainerStateRunning {
```

Comment:
We tried to implement sidecar ourselves with slightly different feature requests - if we kill sidecar and kill pod at the same time, pod status update will be messaged up. it's better to wait for all containers to finish (exited and will not be restarted) before we set KillPod = true